### PR TITLE
T1 (#808): softwareUpdateHandler — resumable downloads to .artifacts/

### PR DESCRIPTION
**Pull request by GOKULRAJ136**
Refer Story #807 
Fixes #808 

## Release Notes

* **New Features**
  * Software updates now support resumable downloads—interrupted transfers can resume from the last point instead of restarting completely. Failed attempts retain partial downloads for subsequent resumption.
  * Disk space is automatically verified before downloads commence to prevent failures due to storage limitations.
  * Enhanced integrity validation ensures downloaded files are complete and uncorrupted.

## Summary by CodeRabbit

* **New Features**
  * Added resumable download support for update artifacts with safe staging/finalization and sidecar metadata.
  * Software update jar handling (v1.3.0+) now records missing/checksum-invalid artifacts without clearing the artifacts area, enabling resume.
  * Added disk-space checks and configurable connection/read timeouts for resumable downloads.

* **Bug Fixes**
  * Improved recovery for stale/invalid resume data (including 206/416 cases), servers ignoring Range, and truncated responses (throws while retaining partials).

* **Tests**
  * Added end-to-end tests covering resume, restart/discard behavior, edge cases, and invalid inputs.

### CODE REVIEW

File: registration/registration-services/src/main/java/io/mosip/registration/update/ResumableDownloader.java [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+        private final File part;
+        private final File meta;
+
+        private Artifact(File dir, String name) {
+            this.dir = dir;
+            this.name = name;
+            this.target = new File(dir, name);
+            this.part = new File(dir, name + PART_SUFFIX);
+            this.meta = new File(dir, name + META_SUFFIX);
+        }
```
[the remark this answers is not in the document]
Comment (GOKULRAJ136) [reply registration/registration-services/src/main/java/io/mosip/registration/update/ResumableDownloader.java:382]: corrected

File: registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateUtil.java [hunk 2]
```diff
@@ -107,6 +110,57 @@ protected static InputStream download(String url) throws RegBaseCheckedException
...
+     */
+    protected static void downloadResumable(String url, String targetDir, String fileName) throws RegBaseCheckedException {
+        try {
+            ResumableDownloader.download(url, targetDir, fileName,
+                    getTimeout(CONNECTION_TIMEOUT, DEFAULT_CONNECTION_TIMEOUT),
+                    getTimeout(READ_TIMEOUT, DEFAULT_READ_TIMEOUT));
+        } catch (IOException e) {
+            LOGGER.error("Failed to download {}", url, e);
+            throw new RegBaseCheckedException("REG-BUILD-005", "Failed to download " + url);
+        }
```
[the remark this answers is not in the document]
Comment (GOKULRAJ136) [reply registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateUtil.java:143]: Corrected

File: registration/registration-services/src/main/java/io/mosip/registration/update/ResumableDownloader.java [hunk 3]
```diff
@@ -0,0 +1,381 @@
+package io.mosip.registration.update;
```
Comment (anushasunkada) [inline registration/registration-services/src/main/java/io/mosip/registration/update/ResumableDownloader.java:6]: Add license text

Comment (GOKULRAJ136) [reply registration/registration-services/src/main/java/io/mosip/registration/update/ResumableDownloader.java:6]: Added license text

File: registration/registration-services/src/main/java/io/mosip/registration/update/ResumableDownloader.java [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+        throw new IOException("Failed to download " + url + " after restart");
+    }
+
+    /**
+     * Performs a single download attempt.
+     *
+     * @return {@code true} if the file was completed and finalized; {@code false} if the partial was
+     *         found stale/unusable and discarded, so the caller should restart from scratch.
+     */
+    private static boolean attemptDownload(String url, Artifact artifact, boolean allowResume,
```
Comment (anushasunkada) [inline registration/registration-services/src/main/java/io/mosip/registration/update/ResumableDownloader.java:103]: @GOKULRAJ136 Instead of writing our implementation for resumable file download, why not use OkHttp3 library which already supports resumable file downloads.

Comment (GOKULRAJ136) [reply registration/registration-services/src/main/java/io/mosip/registration/update/ResumableDownloader.java:103]: @anushasunkada I did evaluate it.

OkHttp3 doesn't provide resumable downloads as a built-in feature. It simply allows you to set the `Range` header, which `HttpURLConnection` already supports. The actual resumable download logic, including persisting the `.part` file, validating with `ETag`/`If-Range` to prevent mixing different versions, handling `200/206/416` responses, verifying `Content-Length`, and performing an atomic finalization, must be implemented manually regardless of which HTTP client is used.

Additionally, OkHttp is not currently a runtime dependency (it is only used by `mockwebserver` for tests), so adopting it would introduce a new runtime dependency. `ResumableDownloader` was intentionally designed to depend only on the JDK and `slf4j` so that T2 can move it into `registration-launcher-common` unchanged and share it with `_launcher.jar`.

File: registration/registration-services/src/test/java/io/mosip/registration/test/update/ResumableDownloaderTest.java [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+
+    @Test(expected = IOException.class)
+    public void ensureSpace_insufficient_throws() throws Exception {
+        File dir = tempDir();
+        try {
+            ResumableDownloader.ensureSpace(dir, Long.MAX_VALUE);
+        } finally {
+            deleteDir(dir);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (GOKULRAJ136) [reply registration/registration-services/src/test/java/io/mosip/registration/test/update/ResumableDownloaderTest.java:264]: Good catch — added ensureSpace_sufficient_doesNotThrow and ensureSpace_nullOrMissingDir_fallsBackToCwd covering the happy path and the null/non-existent-dir fallback.

File: registration/registration-services/src/test/java/io/mosip/registration/test/update/ResumableDownloaderTest.java [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+    public void download_blankFileName_throwsIllegalArgument() throws Exception {
+        ResumableDownloader.download("http://localhost/file", tempDir().getAbsolutePath(), "",
+                CONNECT_TIMEOUT, READ_TIMEOUT);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void download_pathTraversalName_throwsIllegalArgument() throws Exception {
+        ResumableDownloader.download("http://localhost/file", tempDir().getAbsolutePath(), "..\\evil.bin",
+                CONNECT_TIMEOUT, READ_TIMEOUT);
+    }
```
[the remark this answers is not in the document]
Comment (GOKULRAJ136) [reply registration/registration-services/src/test/java/io/mosip/registration/test/update/ResumableDownloaderTest.java:344]: it's minor, should've handled in other places or we can take it later

File: registration/registration-services/src/main/java/io/mosip/registration/update/ResumableDownloader.java [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+                            artifact.name, start, existing);
+                    artifact.discard();
+                    return false;
+                }
+                append = true;
+            } else if (status == HttpURLConnection.HTTP_OK) {          // 200 - fresh body (or If-Range mismatch)
+                append = false;
+                existing = 0L;
+                // Persist the validator so a later interrupted attempt can resume this exact resource.
+                writeValidator(artifact.meta, extractValidator(connection));
```
[the remark this answers is not in the document]
Comment (GOKULRAJ136) [reply registration/registration-services/src/main/java/io/mosip/registration/update/ResumableDownloader.java:144]: resolved

File: registration/registration-services/src/test/java/io/mosip/registration/test/update/ResumableDownloaderTest.java [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+            writeMeta(dir, ETAG);
+
+            ResumableDownloader.download(urlFor(server), dir.getAbsolutePath(), "artifact.bin",
+                    CONNECT_TIMEOUT, READ_TIMEOUT);
+
+            assertArrayEquals(payload, Files.readAllBytes(new File(dir, "artifact.bin").toPath()));
+        } finally {
+            stop(server, dir);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (GOKULRAJ136) [reply registration/registration-services/src/test/java/io/mosip/registration/test/update/ResumableDownloaderTest.java:142]: resolved

File: registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateUtil.java
Comment (anushasunkada) [inline registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateUtil.java:1]: Is clearTempDirectory() used anywhere else?

Comment (GOKULRAJ136) [reply registration/registration-services/src/main/java/io/mosip/registration/update/SoftwareUpdateUtil.java:1]: `clearTempDirectory()` is still in use. `ClientSetupValidator.validateBuildSetup()` (around line 100) invokes it on every non-`LOCAL` startup to clear the `.TEMP/` directory before the checksum verification and patch download flow begins, so it is not dead code.

Additionally, this method and its call site predate this PR (they were introduced in 2022, and the corresponding test was added as part of the coverage PR #761). Since they are outside the scope of **T1**, I have left them unchanged.

**Review by anushasunkada — APPROVED** [review]
